### PR TITLE
Rebranding Verizon Media to Yahoo iOS Adapter.

### PR DESCRIPTION
**Pull request by joshh-devrel**
Rebranding Verizon Media to Yahoo iOS Adapter.

### CODE REVIEW

File: adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooConstants.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+
+/// Yahoo mediation network adapter version.
+static NSString *const _Nonnull GADMAdapterYahooVersion = @"1.14.1.1";
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooConstants.h:18]: Please update to version "1.4.0.1"  - e.g. version 1 of the adapter for Yahoo Mobile SDK v1.4.0

File: adapters/YahooAdapter/README.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
-# Verizon Media Mediation Adapter for Google Mobile Ads SDK for iOS
+# Yahoo Mediation Adapter for Google Mobile Ads SDK for iOS
 
 This is an adapter to be used in conjunction with the Google Mobile Ads SDK.
 For requirements, instructions, and other info, see the
-[Verizon Media Adapter Integration Guide](https://developers.google.com/admob/ios/mediation/verizon-media).
+[Yahoo Adapter Integration Guide](https://developers.google.com/admob/ios/mediation/verizon-media).
```
Comment (thomas-poland) [inline adapters/YahooAdapter/README.md:5]: Currently this link points to the Verizon SDK Integration Guide, obviously not part of this PR, please rebrand the web integration guide as well.

File: adapters/YahooAdapter/Public/Headers/VerizonMediaAdapter.h [hunk 3]
```diff
@@ -16,4 +16,3 @@
 #import <UIKit/UIKit.h>
 
 // Import any public headers below.
-#import <VerizonMediaAdapter/GADMVerizonPrivacy.h>
```
Comment (thomas-poland) [inline adapters/YahooAdapter/Public/Headers/VerizonMediaAdapter.h:19]: This file has the old name, and, is effectively n/a since the privacy files have been removed, probably this file and the module map can be deleted as well; noting that I have yet to verify the adapter for the updated Yahoo SDK privacy APIs fully.

File: adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooBaseClass.m [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  return self;
+}
+
+- (void)dealloc {
+  [self stopBeingDelegate];
+}
+
+#pragma mark - Banner
+
+- (void)getBannerWithSize:(GADAdSize)gadSize {
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooBaseClass.m:74]: I verified banner/inline with our test ads, the banners display and all callbacks are working correctly.

File: adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooBaseClass.m [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+- (void)inlineAd:(nonnull YASInlineAdView *)inlineAd
+           event:(nonnull NSString *)eventId
+          source:(nonnull NSString *)source
+       arguments:(nonnull NSDictionary<NSString *, id> *)arguments {
+  // A generic callback that does currently need an implementation for inline placements.
+}
+
+#pragma mark - Interstital
+
+- (void)getInterstitial {
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooBaseClass.m:179]: I verified interstitial with one of our test ads, the interstitial displays and all callbacks are working correctly.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+      _placementID = credentials[GADMAdapterYahooPosition];
+    }
+    NSString *siteID = credentials[GADMAdapterYahooDCN];
+    GADMAdapterYahooInitializeYASAdsWithSiteID(siteID);
+  }
+
+  return self;
+}
+
+- (void)loadNativeAdWithAdTypes:(nonnull NSArray<GADAdLoaderAdType> *)adTypes
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:59]: I have some questions regarding native ads - while our test ad loads, I don't see click handling as I would expect, I will investigate further.

Please see below for a couple additional comments & questions.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+      YASErrorInfo *error =
+          [(id<YASNativeVideoComponent>)videoComponent prepareView:_yahooVideoPlayerView];
+      if (error) {
+        NSLog(@"[YahooAdapter]: Failed to render mediaView video: %@", error);
+      }
+      return _yahooVideoPlayerView;
+    }
+  }
+
+  component = [_nativeAd component:@"mainImage"];
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:252]: We understood mediaView is for a video view only - please see the Verizon adapter, and, the 'hasVideoContent' method below.   Based on this, lines 252-265 should be deleted.
The mainImage component is handled in the 'images' method above, and it is loading and displaying.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+- (void)didRecordImpression {
+  [_nativeAd fireImpression];
+}
+
+- (void)didRecordClickOnAssetWithName:(GADNativeAssetIdentifier)assetName
+                                 view:(UIView *)view
+                       viewController:(UIViewController *)viewController {
+  [_nativeAd invokeDefaultAction];
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:316]: I do not see this method being called when the ad or its components are clicked.  

So, a question: I have a draft adapter working with Google Mobile Ads SDK v8.9.0, where I do see this method being called, though 'handleUserClicks' is not implemented.  In this PR's adapter using Google Ads v10.0.0,  if I add 'handleUserClicks' and return NO, I still don't see this method being called.   

There are other ways to set up the native ad text labels and image views if we expect the Yahoo SDK to handle user clicks, please see https://sdk.yahooinc.com/yahoo-ads/native.html and the Yahoo Sample App https://github.com/yahoo/YahooMobileSDK-SampleApps-iOS/blob/main/ObjC/SampleApp/NativeContainerView.m, in particular the 'prepareView', 'prepareLabel', and 'prepareButton' APIs.

However, if we don't intend to have Yahoo handle clicks, I would expect this method to be called, perhaps I'm overlooking something.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+- (void)loadNativeAdWithAdTypes:(nonnull NSArray<GADAdLoaderAdType> *)adTypes
+                        options:(nullable NSArray<GADAdLoaderOptions *> *)options {
+  if (![self prepareAdapterForAdRequest]) {
+    return;
+  }
+
+  YASNativePlacementConfig *placementConfig =
+      [[YASNativePlacementConfig alloc] initWithPlacementId:_placementID
+                                            requestMetadata:nil
+                                              nativeAdTypes:@[ @"100", @"simpleImage" ]];
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:68]: For Yahoo Mobile SDK, the nativeAdTypes are 'simpleImage' and 'simpleVideo', please see [the docs here](https://sdk.yahooinc.com/yahoo-ads/native.html#2-create-a-native-ad-placement-configuration)

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+
+- (void)didRenderInView:(UIView *)view
+       clickableAssetViews:(NSDictionary<GADNativeAssetIdentifier, UIView *> *)clickableAssetViews
+    nonclickableAssetViews:
+        (NSDictionary<GADNativeAssetIdentifier, UIView *> *)nonclickableAssetViews
+            viewController:(UIViewController *)viewController {
+  id<YASComponent> iconComponent = [_nativeAd component:@"iconImage"];
+  if ([iconComponent conformsToProtocol:@protocol(YASNativeImageComponent)]) {
+    UIImageView *iconView = (UIImageView *)clickableAssetViews[GADNativeIconAsset];
+    [((id<YASNativeImageComponent>)iconComponent) prepareView:iconView];
```
Comment (lcsabaignat) [inline adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:327]: I see the icon image is handled twice, once here, and once via the 
```
- (GADNativeAdImage *)icon {
  return [self imageForComponent:@"iconImage"];
}
```
Is there a reason for this? 

The icon image ends up having different behaviour from the rest of the native components, where the clicks are handled with the YahooAds, while all other components handle the clicks with the 
didRecordClickOnAssetWithName:view:viewController: call.

File: adapters/YahooAdapter/YahooAdapter/Rewarded/GADMAdapterYahooRewardedAd.m [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  /// An ad event delegate to invoke when ad rendering events occur.
+  __weak id<GADMediationRewardedAdEventDelegate> _adEventDelegate;
+
+  /// Ad Configuration for the ad to be rendered.
+  GADMediationRewardedAdConfiguration *_adConfiguration;
+
+  BOOL _isVideoCompletionEventCalled;
+}
+
+- (void)loadRewardedAdForAdConfiguration:(nonnull GADMediationRewardedAdConfiguration *)adConfig
```
Comment (thomas-poland) [inline adapters/YahooAdapter/YahooAdapter/Rewarded/GADMAdapterYahooRewardedAd.m:45]: I verified RewardedAd with one of our test ads, the video plays and all callbacks are working correctly.

File: adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooConstants.h (hunk 1 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/GADMAdapterYahooConstants.h:18]: Updated.

File: adapters/YahooAdapter/README.md (hunk 2 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/README.md:5]: The mediation developer guide will also be updated.

File: adapters/YahooAdapter/Public/Headers/VerizonMediaAdapter.h (hunk 3 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/Public/Headers/VerizonMediaAdapter.h:19]: Removed from the modulemap.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m (hunk 9 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:68]: Updated.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m (hunk 7 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:252]: My understanding is that without a call to `prepareView`, the `mainImage` asset will not be rendered, which is reflected on my testing.
If I remove lines `252`-`265`, then the image is not rendered.

For reference, I used the example app in https://github.com/googleads/googleads-mobile-ios-examples to verify on my end.

**Comment by joshuajplagonera** [comment]
Hi @thomas-poland 
Responded to your comments. Can you take another look ?

**Comment by thomas-poland** [comment]
@joshuajplagonera ok I tried the latest - as noted, no mainImage, so, indeed we need to use `prepareView`, please see below what I did that works in my test app: do this one place in `imageForComponent` (rather than `didRenderInView` for icon only):

```
- (GADNativeAdImage *)imageForComponent:(NSString *)componentId {
    GADNativeAdImage *GADImage;

    UIView *imageView;
    id<YASNativeImageComponent> imageComponent = (id<YASNativeImageComponent>)[_nativeAd component:componentId];
    if ( imageComponent ) {
        imageView = [[UIImageView alloc] init];
        imageView.contentMode = UIViewContentModeScaleAspectFit;
        [imageComponent prepareView: (UIImageView *) imageView];
        UIImage *image = ((UIImageView *) imageView).image;
        if (image) {
            GADImage = [[GADNativeAdImage alloc] initWithImage:image];
        }
    } 
return GADImage;
}
```

**Comment by joshuajplagonera** [comment]
> @joshuajplagonera ok I tried the latest - as noted, no mainImage, so, indeed we need to use `prepareView`, please see below what I did that works in my test app: do this one place in `imageForComponent` (rather than `didRenderInView` for icon only):
> 
> ```
> - (GADNativeAdImage *)imageForComponent:(NSString *)componentId {
>     GADNativeAdImage *GADImage;
> 
>     UIView *imageView;
>     id<YASNativeImageComponent> imageComponent = (id<YASNativeImageComponent>)[_nativeAd component:componentId];
>     if ( imageComponent ) {
>         imageView = [[UIImageView alloc] init];
>         imageView.contentMode = UIViewContentModeScaleAspectFit;
>         [imageComponent prepareView: (UIImageView *) imageView];
>         UIImage *image = ((UIImageView *) imageView).image;
>         if (image) {
>             GADImage = [[GADNativeAdImage alloc] initWithImage:image];
>         }
>     } 
> return GADImage;
> }
> ```

Hi @thomas-poland 
Thank you for the guidance. I have tried this on my end and confirmed it is working.
I will update the pull-request shortly and let you know.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m (hunk 8 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:303]: This should now be called for all assets.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m (hunk 10 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:314]: This is now removed from `didRenderInView` and the component is now prepared in `imageForComponent`.

File: adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m (hunk 6 above)
Comment (joshuajplagonera) [reply adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m:59]: We only have test placements for the `simpleImage` format. Can you confirm that the adapter also works as expected for `simpleVideo` format?

**Comment by joshuajplagonera** [comment]
Hi @thomas-poland ,
PR should now be updated. Can you take another look ?

**Comment by thomas-poland** [comment]
https://user-images.githubusercontent.com/5540119/228062193-2b76aa9e-d049-4116-ac4e-e7f9837720b7.mov

**Comment by joshuajplagonera** [comment]
Hi @thomas-poland 

Thank you for the guidance. I updated the pull-request based on your comments.
I opted to use `_yasYahooVideoPlayerView` in `-hasVideoContent:` to make sure it is only returning `YES` if the actual player view is set.

Additionally, the sample Ad Unit ID you provided for `simpleVideo` format is not filling on my end.

That said, let us know if this looks good for you and provide approval if so.

**Comment by thomas-poland** [comment]
@joshuajplagonera the changes look correct; I'll be able to test tomorrow (Wed. morning).

**Comment by thomas-poland** [comment]
@joshuajplagonera almost there - 

When testing, for me, `hasVideoContent` is consistently called before `mediaView`, this results in `_yahooVideoPlayerView` is always being nil, so the mainImage displays instead of the video.   By using `self.mediaView` we can be sure the video component has a chance to load - there may be other approaches, e.g. load the videoComponent at some other point before `hasVideoContent` executes. 

For testing the placement for simpleVideo, `ca-app-pub-7866899558273871/7268163816`, please try it having set up siteId `8a809417014d4ce79fa8059b42390154`

**Comment by joshuajplagonera** [comment]
Hi @thomas-poland 

Thank you for the guidance. I have confirmed that video is now rendering properly.
Let us know if the changes look good and provide approval on the Pull-request if there are no more changes needed.

**Review by thomas-poland — APPROVED** [review]
**Approved** Code reviewed, functional test OK on iPhone 13 / iOS 16.0 - banner, interstitial, rewarded, native (simpleImage and simpleVideo) load, render and behave correctly.   Thanks @joshuajplagonera @joshh-devrel
